Invalid and Inactive Participant improvements

patimen opened the pull request:
## Excluding Invalidated Participants entirely
### The Problem
Previous implementation had several key oversights when a participant was invalidated because of invalid inferences or downtime:
1. ActiveParticipants still included them, and gave no indication they had been excluded.
2. They were not removed from all EpochGroup sub-groups.

Additionally, the logic around removal was scattered and hard to understand, which is partly the reason these had not been addressed earlier.

### The Fix
#### ExcludedParticipants list
We add `ExcludedParticipants` to the state. It is keyed by epoch index and participant id. It includes the specific height of the exclusion and the reason for it. At the moment, exclusion is purely for invalidation because of too many invalid inferences. Other reasons are planned in the near future, thus the open-ended nature of the list.

When a participant is initially marked as invalidated for an epoch, they are added to the list.

#### ActiveParticipants response includes ExcludedParticipants
The response now includes the list of excluded participants, which clients will use to filter the Active list. The actual Active list remains the same, as it is important to have a cryptographically traceable ActiveParticipants list and this would become impractical if it also included arbitrary changes during an Epoch. (see related changes in this release).

#### Collateral has a hard dedupe for slashing
Slashing cannot rely on application logic to ensure that a single participant is slashed for a given reason only once per epoch. Given the dire nature of the punishment, we introduce a list in x/collateral that will ensure this at the lowest level, making application logic impossible to result in double punishment.

## Status update refactoring
### The Problem
1. Status update was not uniformly calculated when states changed, thus it would be easy to forget to call the right methods to check status. 
2. The algorithm for invalid status was naive and likely to have both false positives and false negatives, especially during the early phase of an Epoch
3. Inactivity was detected and punished only at the end of an epoch during Claims, leaving inactive participants on the network

### The Fix
#### Unified status calculation
- Status is now calculated alongside `SetParticipant`. Additionally, if a significant state change is detected(from ACTIVE to INVALID for instance), then the procedures for removing the participant are executed.
- Status calculations, for both Invalid and Inactive, have been standardized to use an SPRT algorithm. (see sprt.go and sprt.html in `calculations`.
- Once a participant is moved to INVALID or INACTIVE, it will not be moved back until the next epoch.
- Many governance parameters added to address this.

### Minor additional fixes
#### Participant status reset with new participant
A participant could "re-add" themselves and erase their current epoch stats for the epoch. This could both inadverantly punish good behavior by erasing it or allow masking of bad behavior.

#### Ramping status deprecated
Early plans for using ramping never materialized. For now, this status is superfluous and only adds complication and confusion.

patimen committed:
invalidation improvements

patimen committed:
Some test fixes as well

DimaOrekhovPS commented on inference-chain/x/collateral/module/hooks.go:
```diff
@@ -81,6 +80,7 @@ func (h StakingHooks) BeforeValidatorSlashed(ctx context.Context, valAddr sdk.Va
 
-	_, err := h.k.Slash(sdkCtx, accAddr, fraction)
+	// Tendermint driven slashing is not limited per epoch, so pass in a blank reason
+	_, err := h.k.Slash(sdkCtx, accAddr, fraction, "")
```
but if we always pass the same string, won't we hit "already slashed"?

patimen committed:
Refactor invalidation logic entirely
More description coming

patimen committed:
Fix validator test

patimen committed:
Fix collateral test

patimen committed:
Fixes for more tests
1. getConfig needs to be from the API, not from the file (and it needs to be fixed for snake_case).
2. Add missing parameters to tests

patimen committed:
Add parameter defaults for upgrades, fix tests

0xBECEDA reviewed (approved):
Generally looks good to me, added couple minor comments.

DimaOrekhovPS commented on inference-chain/proto/inference/inference/participant.proto:
```diff
@@ -8,7 +9,7 @@ enum ParticipantStatus {
   INACTIVE = 2;
   INVALID = 3;
-  RAMPING = 4;
+  RAMPING = 4 [deprecated = true];
```
Deprecating it with this update? Lately I saw a message on discord that someone got stuck in the RAMPING status initially, maybe they just didn't get any inference to work their way into a different status

DimaOrekhovPS commented on inference-chain/x/inference/calculations/status.go:
```diff
@@ -0,0 +1,139 @@
+
+	// If we have consecutive failures with a likelihood of less than 1 in a million times, we're assuming bad
+	falsePositiveRate := validationParameters.FalsePositiveRate.ToDecimal()
+	if probabilityOfConsecutiveFailures(falsePositiveRate, newValue.ConsecutiveInvalidInferences).LessThan(decimal.NewFromFloat(0.000001)) {
```
Should we extract 0.000001 into a const?

patimen committed:
Add a test for url updating
And that submit_new_participant doesn't overwrite existing participant

patimen committed:
Merge branch 'upgrade-v0.2.5' of https://github.com/product-science/inference-ignite into jl/invalidation_improvements

# Conflicts:
#	inference-chain/api/inference/inference/query.pulsar.go
#	inference-chain/app/upgrades/v0_2_5/upgrades.go
#	inference-chain/proto/inference/inference/query.proto
#	inference-chain/x/inference/keeper/keeper.go
#	inference-chain/x/inference/types/keys.go
#	inference-chain/x/inference/types/query.pb.go

patimen committed:
Regen proto files

patimen committed:
Fixes for more tests
re-enable the remove for invalid test

patimen committed:
Fix unit tests

patimen commented on inference-chain/x/inference/epochgroup/epoch_group.go:
```diff
@@ -334,13 +334,24 @@ func (eg *EpochGroup) updateMember(ctx context.Context, address string, weight i
 	return err
 }
 
-func (eg *EpochGroup) UpdateMember(ctx context.Context, previousVersion *types.Participant, currentVersion *types.Participant) error {
```
We don't currently have a scenario where we merely edit a member.

patimen committed:
Address PR comments.

patimen committed:
Add quickfailurethreshold for consecutive invalidations

patimen committed:
Fix tests
